site/en/blog/renderingng-data-structures/index.md

```diff
@@ -0,0 +1,689 @@
+If it happens to have multiple effects,
+there may be more than one property tree node in each tree for the same element.
+
+The topology of each tree is like a sparse representation of the DOM.
+For example, if there are three DOM elements with overflow clips,
+then there will be three clip tree nodes,
+and the structure of the clip tree will follow
+the containing-block relationship between the overflow clips.
+There are also links between the trees.
+These links indicate the relative DOM hierarchy,
+and hence order of application, of the nodes.
+For example, if a transform on a DOM element is below another DOM element with a filter,
+then of course the transform applies before the filter.
+
+Each DOM element has a _property tree state_,
+which is a 4-tuple (transform, clip, effect, scroll)
+that indicates the nearest ancestor clip,
+transform and effect tree nodes that take effect on that element.
+This is very convenient, because with this information we know exactly the list of clips,
+transforms, and effects that apply to that element,
+and in which order.
+This tells us where it is on screen and how to draw it.
+
+{% Aside %}
+There are four trees because scrolling only applies to the contained subtree.
+An element is contained by a scroller if it scrolls with it,
+and `position:absolute` and `position: fixed`
+elements often escape ancestor scrolling elements,
+in situations where the scrolling element is not in the containing block chain
+([containing block](https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block),
+recursively) of the fixed- or absolute-positioned element.
+
+However, all other visual effects apply to the entire DOM subtree.
+Because of this mismatch,
+the topology of the clipping and scrolling aspects of property trees
+is sometimes quite different than for visual effects.
+This is also why most visual effects induce a containing block for all descendants
+([example](https://github.com/w3c/fxtf-drafts/issues/11)).
+The structure of property trees in all their glory is quite complicated;
+see the long code comments
+[here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/paint/object_paint_properties.h;l=96)
+for lots of detail on this subject.
+{% endAside %}
+
+### Example
+
+([source](https://output.jsbin.com/goqojuz/quiet))
+
+```html
+<html>
+  <div style="overflow: scroll; width: 100px; height: 100px;">
+    <iframe style="filter: blur(3px); transform: rotateZ(1deg); width: 100px; height: 300px"
+  id="one" srcdoc="iframe one"></iframe>
+  </div>
+  <iframe style="top:200px; transform: scale(1.1) translateX(200px)" id=two
+      srcdoc="iframe two"></iframe>
+</html>
+```
+
+For the preceding example (which is slightly different than the one in the introduction),
+here are the key elements of the property trees generated:
+
+{% Img src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/ahpRNDTL3cijYyzazyz0.jpg",
```

Comment:
This image has a black line on the RHS. Why?

@una

Reply:
https://deploy-preview-1184--developer-chrome-com.netlify.app/blog/renderingng-data-structures/

Reply:
It's on the image file, I've removed it and reuploaded the image so will be fixed when I push these changes.